### PR TITLE
Locker API

### PR DESCRIPTION
**Pull request by VALERA771**

### CODE REVIEW

File: Exiled.Events/Patches/Events/Map/SpawningItemInLocker.cs [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+using MapGeneration.Distributors;
+
+namespace Exiled.Events.Patches.Events.Map
+{
+    [HarmonyPatch(typeof(LockerChamber), nameof(LockerChamber.SpawnItem))]
+    internal class SpawningItemInLocker
+    {
+
+    }
+}
```
Comment (louis1706) [inline Exiled.Events/Patches/Events/Map/SpawningItemInLocker.cs:11]: there something missing :trollface:

**Review by louis1706 — CHANGES_REQUESTED** [review]

File: Exiled.Events/Patches/Events/Map/SpawningItemInLocker.cs (hunk 1 above)
Comment (VALERA771) [reply Exiled.Events/Patches/Events/Map/SpawningItemInLocker.cs:11]: Yes, so it's still draft

**Comment by VALERA771** [comment]
Transpiler is tested and not causing errors on server startup

File: Exiled.API/Features/Lockers/Chamber.cs [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+        /// Gets or sets a value indicating whether or not items should be spawned as soon as they are initialised.
+        /// </summary>
+        public bool InitiallySpawn
+        {
+            get => Base._spawnOnFirstChamberOpening;
+            set => Base._spawnOnFirstChamberOpening = value;
+        }
+
+        /// <summary>
+        /// Gets or sets a time before player can again interact chamber.
```
Comment (Thundermaker300) [inline Exiled.API/Features/Lockers/Chamber.cs:142]:
```suggestion
        /// Gets or sets the amount of time before a player can interact with the chamber again.
```

File: Exiled.API/Features/Lockers/Chamber.cs [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+        /// </summary>
+        public float Cooldown
+        {
+            get => Base._targetCooldown;
+            set => Base._targetCooldown = value;
+        }
+
+        /// <summary>
+        /// Gets a value indicating whether or not player can interact with chamber now.
```
Comment (Thundermaker300) [inline Exiled.API/Features/Lockers/Chamber.cs:151]:
```suggestion
        /// Gets a value indicating whether or not the chamber can currently be interacted with.
```

**Review by Thundermaker300 — CHANGES_REQUESTED** [review]

File: Exiled.Events/EventArgs/Player/InteractingLockerEventArgs.cs [hunk 4]
```diff
@@ -39,21 +39,21 @@ public class InteractingLockerEventArgs : IPlayerEvent, IDeniableEvent
...
         ///     Gets the <see cref="MapGeneration.Distributors.Locker" /> instance.
         /// </summary>
-        public Locker Locker { get; }
+        public API.Features.Lockers.Locker Locker { get; }
 
         /// <summary>
         ///     Gets the interacting chamber.
         /// </summary>
-        public LockerChamber Chamber { get; }
+        public API.Features.Lockers.Chamber Chamber { get; }
```
Comment (louis1706) [inline Exiled.Events/EventArgs/Player/InteractingLockerEventArgs.cs:56]: No breaking change allowed now

**Review by louis1706 — CHANGES_REQUESTED** [review]

File: Exiled.Events/Patches/Events/Map/SpawningItemInLocker.cs [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+                    // if (!ev.IsAllowed)
+                    //  return;
+                    new(OpCodes.Callvirt, PropertyGetter(typeof(SpawningItemInLockerEventArgs), nameof(SpawningItemInLockerEventArgs.IsAllowed))),
+                    new(OpCodes.Brtrue_S, continueLabel),
+
+                    new(OpCodes.Ret),
+
+                    // if (ipb.Info.Serial != ev.Pickup.Serial)
+                    //      Object.Destroy(ipb);
+                    //      ipb = ev.Pickup.Base;
```
Comment (IRacle1) [inline Exiled.Events/Patches/Events/Map/SpawningItemInLocker.cs:84]: i think better to check by `Base` itself, `ipb != ev.Pickup.Base`, not by serial, or i cant be do by any reason?

**Review by IRacle1 — COMMENTED** [review]
also, that pr has conflicts with current dev, i made `SpawningItemInLocker` event and `LockerList` patch too. if you'll have trouble resolving the merge conflicts/have questions about current locker system on dev - contact me

File: Exiled.API/Features/Lockers/Locker.cs [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+        /// <param name="predicate">Predicate to match.</param>
+        /// <returns>List of <see cref="Locker"/> which matches predicate.</returns>
+        public static IEnumerable<Locker> Get(System.Func<Locker, bool> predicate) => List.Where(predicate);
+
+        /// <summary>
+        /// Interacts with a specific chamber.
+        /// </summary>
+        /// <param name="chamber">Chamber. If <see langword="null"/>, will interact with random.</param>
+        /// <param name="player">Player who interacts.</param>
+        public void Interact(Chamber chamber = null, Player player = null) => Base.ServerInteract(player?.ReferenceHub, (byte)(chamber == null ? Random.Range(0, Chambers.Count + 1) : Chambers.ToList().IndexOf(chamber)));
```
Comment (louis1706) [inline Exiled.API/Features/Lockers/Locker.cs:89]: I don't thinks that the right code : Random.Range(0, Chambers.Count + 1)

File: Exiled.API/Features/Lockers/Chamber.cs [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+
+        /// <summary>
+        /// Gets a value indicating whether or not the chamber can currently be interacted with.
+        /// </summary>
+        public bool CanInteract => Base.CanInteract;
+
+        /// <summary>
+        /// Gets the locker where this chamber is located.
+        /// </summary>
+        public Locker Locker => Locker.Get(x => x.Chambers.Contains(this)).FirstOrDefault();
```
Comment (louis1706) [inline Exiled.API/Features/Lockers/Chamber.cs:158]: i have some doubt if this is really the best way to get the Locker.

Comment (VALERA771) [reply Exiled.API/Features/Lockers/Chamber.cs:158]: Idk really. Couldn't find better way

File: Exiled.API/Features/Lockers/Locker.cs (hunk 6 above)
Comment (louis1706) [reply Exiled.API/Features/Lockers/Locker.cs:89]: how you want it's Id be greater than the number of chamber it's should - 1

**Review by louis1706 — CHANGES_REQUESTED** [review]
Required fix ```[2023-12-04 15:49:36.643 +01:00] [ERROR] [Exiled.Events] Patching by attributes failed!
                                 HarmonyLib.HarmonyException: Patching exception in method virtual System.Void MapGeneration.Distributors.Locker::Start() ---> System.Reflection.AmbiguousMatchException: Ambiguous match in Harmony patch for Exiled.API.Features.Lockers.Locker:Get ---> System.Reflection.AmbiguousMatchException: Ambiguous match found.
                                   at System.RuntimeType.GetMethodImplCommon (System.String name, System.Int32 genericParameterCount, System.Reflection.BindingFlags bindingAttr, System.Reflection.Binder binder, System.Reflection.CallingConventions callConv, System.Type[] types, System.Reflection.ParameterModifier[] modifiers) [0x00050] in <75633565436c42f0a6426b33f0132ade>:0
                                   at System.RuntimeType.GetMethodImpl (System.String name, System.Reflection.BindingFlags bindingAttr, System.Reflection.Binder binder, System.Reflection.CallingConventions callConv, System.Type[] types, System.Reflection.ParameterModifier[] modifiers) [0x00000] in <75633565436c42f0a6426b33f0132ade>:0
                                   at System.Type.GetMethod (System.String name, System.Reflection.BindingFlags bindingAttr) [0x0000e] in <75633565436c42f0a6426b33f0132ade>:0
                                   at HarmonyLib.AccessTools+<>c__DisplayClass27_0.<Method>b__0 (System.Type t) [0x00000] in <0020ee40255b4dc7a0eb8b2a76225b38>:0
                                   at HarmonyLib.AccessTools.FindIncludingBaseTypes[T] (System.Type type, System.Func`2[T,TResult] func) [0x00000] in <0020ee40255b4dc7a0eb8b2a76225b38>:0
                                   at HarmonyLib.AccessTools.Method (System.Type type, System.String name, System.Type[] parameters, System.Type[] generics) [0x0004b] in <0020ee40255b4dc7a0eb8b2a76225b38>:0
                                    --- End of inner exception stack trace ---
                                   at HarmonyLib.AccessTools.Method (System.Type type, System.String name, System.Type[] parameters, System.Type[] generics) [0x0008e] in <0020ee40255b4dc7a0eb8b2a76225b38>:0
                                   at Exiled.Events.Patches.Generic.LockerList+<Transpiler>d__0.MoveNext () [0x000ae] in <a7b6359813344257ae1ec1e42d708bef>:0
                                   at System.Collections.Generic.List`1[T].AddEnumerable (System.Collections.Generic.IEnumerable`1[T] enumerable) [0x00059] in <75633565436c42f0a6426b33f0132ade>:0
                                   at System.Collections.Generic.List`1[T]..ctor (System.Collections.Generic.IEnumerable`1[T] collection) [0x00062] in <75633565436c42f0a6426b33f0132ade>:0
                                   at System.Linq.Enumerable.ToList[TSource] (System.Collections.Generic.IEnumerable`1[T] source) [0x00018] in <1c318258bf0843289b0e2cbe692fee39>:0
                                   at HarmonyLib.CodeTranspiler.GetResult (System.Reflection.Emit.ILGenerator generator, System.Reflection.MethodBase method) [0x00051] in <0020ee40255b4dc7a0eb8b2a76225b38>:0
                                   at HarmonyLib.MethodBodyReader.FinalizeILCodes (HarmonyLib.Emitter emitter, System.Collections.Generic.List`1[T] transpilers, System.Collections.Generic.List`1[T] endLabels, System.Boolean& hasReturnCode) [0x0014a] in <0020ee40255b4dc7a0eb8b2a76225b38>:0
                                   at HarmonyLib.MethodCopier.Finalize (HarmonyLib.Emitter emitter, System.Collections.Generic.List`1[T] endLabels, System.Boolean& hasReturnCode) [0x00000] in <0020ee40255b4dc7a0eb8b2a76225b38>:0
                                   at HarmonyLib.MethodPatcher.CreateReplacement (System.Collections.Generic.Dictionary`2[System.Int32,HarmonyLib.CodeInstruction]& finalInstructions) [0x002be] in <0020ee40255b4dc7a0eb8b2a76225b38>:0
                                   at HarmonyLib.PatchFunctions.UpdateWrapper (System.Reflection.MethodBase original, HarmonyLib.PatchInfo patchInfo) [0x00057] in <0020ee40255b4dc7a0eb8b2a76225b38>:0
                                   at HarmonyLib.PatchClassProcessor.ProcessPatchJob (HarmonyLib.PatchJobs`1+Job[T] job) [0x000bb] in <0020ee40255b4dc7a0eb8b2a76225b38>:0
                                    --- End of inner exception stack trace ---
                                   at HarmonyLib.PatchClassProcessor.ReportException (System.Exception exception, System.Reflection.MethodBase original) [0x0010f] in <0020ee40255b4dc7a0eb8b2a76225b38>:0
                                   at HarmonyLib.PatchClassProcessor.Patch () [0x00093] in <0020ee40255b4dc7a0eb8b2a76225b38>:0
                                   at Exiled.Events.Features.Patcher.PatchAll (System.Boolean includeEvents, System.Int32& failedPatch) [0x0005b] in <a7b6359813344257ae1ec1e42d708bef>:0```

File: Exiled.API/Features/Map.cs [hunk 8]
```diff
@@ -248,8 +250,15 @@ public static void ResetLightsColor()
...
         /// </summary>
         /// <returns><see cref="Locker"/> object.</returns>
+        [Obsolete("Use GetRandomExiledLocker() instead.")]
         public static Locker GetRandomLocker() => Lockers.GetRandomValue();
 
+        /// <summary>
+        /// Gets a random <see cref="Features.Lockers.Locker"/>.
+        /// </summary>
+        /// <returns><see cref="Features.Lockers.Locker"/> object.</returns>
+        public static Lockers.Locker GetRandomExiledLocker() => Features.Lockers.Locker.List.GetRandomValue();
```
Comment (louis1706) [inline Exiled.API/Features/Map.cs:260]: That an horrible name

**Review by louis1706 — CHANGES_REQUESTED** [review]

File: Exiled.API/Features/Map.cs (hunk 8 above)
Comment (VALERA771) [reply Exiled.API/Features/Map.cs:260]: Any suggestions?

Comment (louis1706) [reply Exiled.API/Features/Map.cs:260]: > Any suggestions?

Make breaking change and just call it GetRandomLocker

Comment (VALERA771) [reply Exiled.API/Features/Map.cs:260]: So can I make everywhere breaking changes? (Where it's not bad)

Comment (louis1706) [reply Exiled.API/Features/Map.cs:260]: Yeah

**Review by louis1706 — APPROVED** [review]

**Review by NaoUnderscore — CHANGES_REQUESTED** [review]
Fix conflicts

**Review by louis1706 — APPROVED** [review]

**Review by louis1706 — APPROVED** [review]

**Review by NaoUnderscore — CHANGES_REQUESTED** [review]
Fix build errors

File: Exiled.API/Features/Lockers/Chamber.cs [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+        /// Gets or sets required permissions to open this chamber.
+        /// </summary>
+        public KeycardPermissions RequiredPermissions
+        {
+            get => (KeycardPermissions)Base.RequiredPermissions;
+            set => Base.RequiredPermissions = (Interactables.Interobjects.DoorUtils.KeycardPermissions)value;
+        }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether or not multiple spawnpoint system should be used.
```
Comment (NaoUnderscore) [inline Exiled.API/Features/Lockers/Chamber.cs:114]:
```suggestion
        /// Gets or sets a value indicating whether multiple spawn points  should be used.
```

File: Exiled.API/Features/Lockers/Chamber.cs [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+            {
+                Base._content.Clear();
+
+                foreach (Pickup pickup in value)
+                    Base._content.Add(pickup.Base);
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets the list with all spawnpoints.
```
Comment (NaoUnderscore) [inline Exiled.API/Features/Lockers/Chamber.cs:84]:
```suggestion
        /// Gets or sets all spawn points.
```

File: Exiled.API/Features/Lockers/Chamber.cs [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+            {
+                Base._toBeSpawned.Clear();
+
+                foreach (Pickup pickup in value)
+                    Base._toBeSpawned.Add(pickup.Base);
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets the list with all pickups in chamber.
```
Comment (NaoUnderscore) [inline Exiled.API/Features/Lockers/Chamber.cs:69]:
```suggestion
        /// Gets or sets all pickups in the chamber.
```

File: Exiled.API/Features/Lockers/Chamber.cs [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+        /// <summary>
+        /// Gets the list with all chambers.
+        /// </summary>
+        public static new IReadOnlyCollection<Chamber> List => Chambers.Values;
+
+        /// <inheritdoc/>
+        public LockerChamber Base { get; }
+
+        /// <summary>
+        /// Gets or sets a list with all pickups that should be spawned when door will be opened first time.
```
Comment (NaoUnderscore) [inline Exiled.API/Features/Lockers/Chamber.cs:54]:
```suggestion
        /// Gets or sets all pickups that should be spawned when the door is initially opened.
```

File: Exiled.API/Features/Lockers/Chamber.cs [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+            : base(chamber.gameObject)
+        {
+            Base = chamber;
+            Locker = locker;
+
+            Chambers.Add(chamber, this);
+        }
+
+        /// <summary>
+        /// Gets the list with all chambers.
```
Comment (NaoUnderscore) [inline Exiled.API/Features/Lockers/Chamber.cs:46]:
```suggestion
        /// Gets all chambers.
```

File: Exiled.API/Features/Lockers/Chamber.cs [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+        /// Will be used if <see cref="UseMultipleSpawnpoints"/> is <see langword="true"/>.
+        /// </remarks>
+        public IEnumerable<Transform> Spawnpoints
+        {
+            get => Base._spawnpoints;
+            set => Base._spawnpoints = value.ToArray();
+        }
+
+        /// <summary>
+        /// Gets or sets the list with <see cref="ItemType"/> which can be spawned in this chamber.
```
Comment (NaoUnderscore) [inline Exiled.API/Features/Lockers/Chamber.cs:96]:
```suggestion
        /// Gets or sets all the acceptable items which can be spawned in this chamber.
```

File: Exiled.API/Features/Lockers/Chamber.cs [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+                foreach (Pickup pickup in value)
+                    Base._content.Add(pickup.Base);
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets the list with all spawnpoints.
+        /// </summary>
+        /// <remarks>
+        /// Will be used if <see cref="UseMultipleSpawnpoints"/> is <see langword="true"/>.
```
Comment (NaoUnderscore) [inline Exiled.API/Features/Lockers/Chamber.cs:87]:
```suggestion
        /// Used if <see cref="UseMultipleSpawnpoints"/> is set to <see langword="true"/>.
```

File: Exiled.API/Features/Lockers/Chamber.cs [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+        /// If true, spawnpoint system will use <see cref="Spawnpoints"/>. Otherwise, <see cref="Spawnpoint"/>.
+        /// </remarks>
+        public bool UseMultipleSpawnpoints
+        {
+            get => Base._useMultipleSpawnpoints;
+            set => Base._useMultipleSpawnpoints = value;
+        }
+
+        /// <summary>
+        /// Gets or sets a spawnpoint for items in locker chamber.
```
Comment (NaoUnderscore) [inline Exiled.API/Features/Lockers/Chamber.cs:126]:
```suggestion
        /// Gets or sets a spawn point for the items in the chamber.
```

File: Exiled.API/Features/Lockers/Chamber.cs [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+        {
+            get => (KeycardPermissions)Base.RequiredPermissions;
+            set => Base.RequiredPermissions = (Interactables.Interobjects.DoorUtils.KeycardPermissions)value;
+        }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether or not multiple spawnpoint system should be used.
+        /// </summary>
+        /// <remarks>
+        /// If true, spawnpoint system will use <see cref="Spawnpoints"/>. Otherwise, <see cref="Spawnpoint"/>.
```
Comment (NaoUnderscore) [inline Exiled.API/Features/Lockers/Chamber.cs:117]:
```suggestion
        /// If <see langword="true"/>, <see cref="Spawnpoints"/> will be used over <see cref="Spawnpoint"/>.
```

File: Exiled.API/Features/Lockers/Chamber.cs [hunk 18]
```diff
@@ -0,0 +1,185 @@
...
+        {
+            get => Base._useMultipleSpawnpoints;
+            set => Base._useMultipleSpawnpoints = value;
+        }
+
+        /// <summary>
+        /// Gets or sets a spawnpoint for items in locker chamber.
+        /// </summary>
+        /// <remarks>
+        /// Will be used if <see cref="UseMultipleSpawnpoints"/> is <see langword="false"/>.
```
Comment (NaoUnderscore) [inline Exiled.API/Features/Lockers/Chamber.cs:129]:
```suggestion
        /// Used if <see cref="UseMultipleSpawnpoints"/> is set to <see langword="false"/>.
```

File: Exiled.API/Features/Lockers/Chamber.cs [hunk 19]
```diff
@@ -0,0 +1,185 @@
...
+        /// </remarks>
+        public Transform Spawnpoint
+        {
+            get => Base._spawnpoint;
+            set => Base._spawnpoint = value;
+        }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether or not items should be spawned as soon as they are initialised.
```
Comment (NaoUnderscore) [inline Exiled.API/Features/Lockers/Chamber.cs:138]:
```suggestion
        /// Gets or sets a value indicating whether or not items should be spawned as soon as they are initialized.
```

File: Exiled.API/Features/Lockers/Chamber.cs [hunk 20]
```diff
@@ -0,0 +1,185 @@
...
+        }
+
+        /// <summary>
+        /// Gets the <see cref="Stopwatch"/> of current cooldown.
+        /// </summary>
+        /// <remarks>Used in <see cref="CanInteract"/> check.</remarks>
+        public Stopwatch CurrentCooldown => Base._stopwatch;
+
+        /// <summary>
+        /// Gets a value indicating whether or not the chamber can currently be interacted with.
```
Comment (NaoUnderscore) [inline Exiled.API/Features/Lockers/Chamber.cs:162]:
```suggestion
        /// Gets a value indicating whether the chamber is interactable.
```

File: Exiled.API/Features/Lockers/Chamber.cs [hunk 21]
```diff
@@ -0,0 +1,185 @@
...
+
+        /// <summary>
+        /// Spawns a specified item from <see cref="AcceptableTypes"/>.
+        /// </summary>
+        /// <param name="type"><see cref="ItemType"/> from <see cref="AcceptableTypes"/>.</param>
+        /// <param name="amount">Amount of items that should be spawned.</param>
+        public void SpawnItem(ItemType type, int amount) => Base.SpawnItem(type, amount);
+
+        /// <summary>
+        /// Gets the chamber by it's <see cref="LockerChamber"/>.
```
Comment (NaoUnderscore) [inline Exiled.API/Features/Lockers/Chamber.cs:179]:
```suggestion
        /// Gets the chamber by its <see cref="LockerChamber"/>.
```

File: Exiled.API/Features/Lockers/Locker.cs [hunk 22]
```diff
@@ -0,0 +1,112 @@
...
+            {
+                PedestalScpLocker psl => new PedestalLocker(psl),
+                _ => new Locker(locker)
+            };
+
+        /// <summary>
+        /// Gets the list of <see cref="Locker"/> which matches predicate.
+        /// </summary>
+        /// <param name="predicate">Predicate to match.</param>
+        /// <returns>List of <see cref="Locker"/> which matches predicate.</returns>
```
Comment (NaoUnderscore) [inline Exiled.API/Features/Lockers/Locker.cs:91]:
```suggestion
        /// <returns>All <see cref="Locker"/> instances matching the predicate.</returns>
```

File: Exiled.API/Features/Lockers/Locker.cs [hunk 23]
```diff
@@ -0,0 +1,112 @@
...
+        public static IEnumerable<Locker> Get(System.Func<Locker, bool> predicate) => List.Where(predicate);
+
+        /// <summary>
+        /// Interacts with a specific chamber.
+        /// </summary>
+        /// <param name="chamber">Chamber. If <see langword="null"/>, will interact with random.</param>
```
Comment (NaoUnderscore) [inline Exiled.API/Features/Lockers/Locker.cs:97]:
```suggestion
        /// <param name="chamber">If <see langword="null"/>, the interaction will be randomized.</param>
```

**Review by NaoUnderscore — CHANGES_REQUESTED** [review]

**Pull request merged**